### ISSUE TITLE
copy_string error

### ISSUE THREAD

**Issue by hailihu** [davidfrantz/force#37]
For a Landsat scene (LC81210622020202LGN00), force-l2ps exits  with error "cannot copy, string too long.". This happens at line 699 of src/lower-level/gas-ll.c:
 `copy_string(source, NPOW_02, tokenptr)`
because apparently `tokenptr` contains a too long string, printing it gives "MYD". This happens when reading a daily `WVP_***.txt` file where "MYD" is in the last column. Perhaps there is no end-of-string character for `tokenptr`? Anyways, replacing line 699 with:
 `strncpy(source, tokenptr, NPOW_02)`
makes the preprocessing succeed.

**Comment by davidfrantz** [comment]
Hi Haili,

thanks for your suggestion. I really appreciate this.

But this is already fixed now with ``FORCE 3.5.2``. 

I would however suggest you not using your proposed fix as this is insecure (have a read [here](https://security.web.cern.ch/recommendations/en/codetools/c.shtml)), and actually fails in this case.

The problem here is, that I forgot to replace the EOL character in the reading funtion, i.e. t``okenptr = "MOD\n"``, should actually be ``tokenptr = "MOD\0"``

This means that the copied string won't be null-terminated, which might cause other problems.

The copy_string function already did as you suggest, but adds a test whether the string is null-terminated after copying. As we were copying 4 (``tokenptr = "MOD\0n``) instead of 3 characters (``tokenptr = "MOD\0"``), there was no space for the terminating NULL, and the function fails safely:

``` bash
void copy_string(char *dst, size_t size, const char *src){

  strncpy(dst, src, size);
  if (dst[size-1] != '\0'){
    printf("cannot copy, string too long:\n%s\n", src);
    exit(1);
  }

  return;
}
```

The fix is to include this line here in ``src/lower-level/gas-ll.c`` in l. 682:

``buffer[strcspn(buffer, "\r\n#")] = 0;``

This even makes it safe to use a file with Windows EOL.


Cheers,

David

**Comment by hailihu** [comment]
Thanks David,
I tested 3.5.2 and it works now. It wasn't really a fix I proposed, just I workaround so I could identify the source of error and continue processing. Good you have a solid fix now, thanks!
Cheers, Haili
